kernel/src/main/java/org/kframework/parser/concrete2kore/disambiguation/KAppToTermConsVisitor.java

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2015-2019 K Team. All Rights Reserved.
+package org.kframework.parser.concrete2kore.disambiguation;
+
+import com.google.common.collect.Sets;
+import org.kframework.builtin.Sorts;
+import org.kframework.definition.Module;
+import org.kframework.definition.Production;
+import org.kframework.parser.Ambiguity;
+import org.kframework.parser.Constant;
+import org.kframework.parser.SetsTransformerWithErrors;
+import org.kframework.parser.Term;
+import org.kframework.parser.TermCons;
+import org.kframework.utils.StringUtil;
+import org.kframework.utils.errorsystem.KEMException;
+import org.pcollections.ConsPStack;
+import org.pcollections.PStack;
+import scala.collection.immutable.Set$;
+import scala.util.Either;
+import scala.util.Left;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.kframework.Collections.*;
+import static org.kframework.kore.KORE.*;
+
+
+/**
+ * Transform the KApps found in a term into the corresponding TermCons so type checking and
+ * variable type inference takes place correctly. Must be applied between type inference and
+ * priority filter.
+ */
+public class KAppToTermConsVisitor extends SetsTransformerWithErrors<KEMException> {
+
+    private final Module mod;
+    public KAppToTermConsVisitor(Module mod) {
+        super();
+        this.mod = mod;
+    }
+
+    @Override
+    public Either<java.util.Set<KEMException>, Term> apply(TermCons tc) {
+        assert tc.production() != null : this.getClass() + ":" + " production not found." + tc;
+        if (tc.production().klabel().isDefined() && tc.production().klabel().get().name().equals("#KApply")) {
+            if (!(tc.items().get(1) instanceof Constant) || !((Constant) tc.items().get(1)).production().sort().equals(Sorts.KLabel()))
```

Comment:
Let's use TermCons.get everywhere so we don't get confused by the reversed indices